contents/performance/Performance.md

```diff
@@ -0,0 +1,102 @@
+# Introduction to Performance Optimization
+
+Author: [Ian Teo](https://github.com/IanTeo)
+
+## Overview
+
+Computer performance can be defined as the rate of work accomplished by a computer system. Even if execution time is not important for a particular application, it may be important to reduce CPU cycles so as to consume less power; from applications running in small battery operated devices to huge data centres.
+
+`Premature optimization is the root of all evil.`<br>
+I think many people have heard of this quote from Donald Knuth before. This quote is actually misinterpreted frequently, because of the lack of context. Here is the full quote:<br>
+`We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.`<br>
+You can find a nice explaination of the quote [here](http://wiki.c2.com/?PrematureOptimization).
+
+This is not a guide on how to optimize that 3%, but rather, to teach you standard optimization techniques that you should apply in all of your code, so that you do not create sub-optimal codes (premature pessimization). If you want to find out more about optimizing that 3%, you can find more at [Other Resources](#other-resources) below.
+
+The 2 techniques you should know are:
+
+* Using appropriate Data Structures and Algorithms
+* Using Cache Friendly codes
+
+## Using Appropriate Data Structures and Algorithms
+
+Using appropriate Data Structures and Algorithms can improve the execution speed of your program trememdously. However, it requires time to learn and understand the nuances of each Data Structure and Algorithm and when to use them.
+
+A good example of how Data Structures and Algorithms can improve execution time is with [binary search](https://www.tutorialspoint.com/data_structures_algorithms/binary_search_algorithm.htm). When finding a particular element in a list, you need to search through the entire list. If it has a million entries, every find will require you to look through **1,000,000** entries. However, if you have a [sorted](https://www.tutorialspoint.com/data_structures_algorithms/sorting_algorithms.htm) list, you can use binary search to reduce the number of entries you have to look through to around **20**!
+
+If you are interested in learning more about Data Structures and Algorithms, you can learn more from the following resources:
+
+* [Geeksforgeeks.org](http://www.geeksforgeeks.org/data-structures/)
+* [Introduction to Algorithms (Book)](https://www.amazon.com/Introduction-Algorithms-3rd-MIT-Press/dp/0262033844)
+* [MIT online course](http://courses.csail.mit.edu/6.851/spring12/lectures/)
+* [topcoder.com](https://www.topcoder.com/community/data-science/data-science-tutorials/)
```

Comment:
Give a brief explanation of what the reader can expect to gain from each resource?